README.md
# DEVIL

A python package that exploits PyTorch and Pyro to perform fast and scalable single cell differential expression analysis

![](logo.png)

## Installation

```bash
$ pip install pydevil
```

## Usage

- TODO



## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`pydevil` was created by Salvatore Milite. It is licensed under the terms of the MIT license.

## Credits

`pydevil` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (26/04/2023)

- First release of `pydevil`!

src/pydevil/tests.py
import numpy as np
from scipy.stats import norm
from statsmodels.stats.multitest import fdrcorrection
import pandas as pd



def test_posterior_null(inference_res,contrast, alpha = 0.05):
    mu_test = (inference_res["params"]["beta"] * contrast.reshape([-1,1])).sum(axis = 0)
    
    variance_term = (inference_res["params"]["variance"].diagonal(0,2,1) * contrast**2).sum(axis = 1)

    covariance_term = 0
    if np.where(contrast > 0)[0] > 1:
        total_terms = len(contrast)
        for i in range(1, total_terms):
            for j in range(0, i - 1):
                covariance_term += contrast[i] * contrast[j] * inference_res["params"]["variance"][:,i,j]

    total_variance = variance_term + covariance_term

    p_value = (1 - norm.cdf(np.abs(mu_test), scale = np.sqrt(total_variance)))
    is_significant, p_value_adj = fdrcorrection(p_value, alpha=alpha)

    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : mu_test,
         "p_value" : p_value, 
         "p_value_adj" : p_value_adj, 
         "is_significant" : is_significant })
    return ret_df

def posterior_CI(inference_res,contrast, credible_mass = 0.95):
    mu_test = (inference_res["params"]["beta"] * contrast.reshape([-1,1])).sum(axis = 0)
    
    variance_term = (inference_res["params"]["variance"].diagonal(0,2,1) * contrast**2).sum(axis = 1)

    covariance_term = 0
    if np.where(contrast > 0)[0] > 1:
        total_terms = len(contrast)
        for i in range(1, total_terms):
            for j in range(0, i - 1):
                covariance_term += contrast[i] * contrast[j] * inference_res["params"]["variance"][:,i,j]

    total_variance = variance_term + covariance_term
    lower = (1 - credible_mass) / 2 
    upper = (1 + credible_mass) / 2     
    interval = np.array([norm.ppf(lower,loc = mu_test ,scale = np.sqrt(total_variance)),norm.ppf(upper,loc = mu_test ,scale = np.sqrt(total_variance))])
    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : mu_test,
         "CI_low" : interval[0,:], 
         "CI_high": interval[1,:], 
         "is_zero_in" : (interval[0,:] < 0) & (interval[1,:] > 0) })
    return ret_df

def test_posterior_ROPE(inference_res,contrast, LFC = 0.5):
    mu_test = (inference_res["params"]["beta"] * contrast.reshape([-1,1])).sum(axis = 0)
    
    variance_term = (inference_res["params"]["variance"].diagonal(0,2,1) * contrast**2).sum(axis = 1)

    covariance_term = 0
    if np.where(contrast > 0)[0] > 1:
        total_terms = len(contrast)
        for i in range(1, total_terms):
            for j in range(0, i - 1):
                covariance_term += contrast[i] * contrast[j] * inference_res["params"]["variance"][:,i,j]

    total_variance = variance_term + covariance_term

    ROPE = norm.cdf(LFC,loc = np.abs(mu_test), scale = np.sqrt(total_variance)) - norm.cdf(-LFC,loc = np.abs(mu_test), scale = np.sqrt(total_variance))

    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : mu_test,
         "ROPE" : ROPE
    })
    return ret_df


def test_posterior_null_HMC(inference_res,contrast, alpha = 0.05):
    beta_stacked = res_de_HMC["params"]["beta"]
    mu_test = (beta_stacked * contrast.reshape([1,1,-1])).sum(axis = 2)
    var_test = np.std(mu_test, axis = 0)
    mu_test = np.mean(mu_test, axis = 0)

    p_value = (1 - norm.cdf(np.abs(mu_test), scale = var_test))
    is_significant, p_value_adj = fdrcorrection(p_value, alpha=alpha)

    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : mu_test,
         "p_value" : p_value, 
         "p_value_adj" : p_value_adj, 
         "is_significant" : is_significant })
    return ret_df

def test_posterior_ROPE_HMC(inference_res,contrast, LFC = 0.5):
    
    beta_stacked = inference_res["params"]["beta"]
    mu_test = (beta_stacked * contrast.reshape([1,1,-1])).sum(axis = 2)
    ROPE = np.sum(np.abs(mu_test) < LFC,  axis = 0)/mu_test.shape[0]


    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : np.mean(mu_test, axis = 0),
         "ROPE" : ROPE 
    })
    return ret_df

def posterior_CI_HMC(inference_res,contrast, credible_mass = 0.95, CI_type = "HPDI"):
    beta_stacked = inference_res["params"]["beta"]
    mu_test = (beta_stacked * contrast.reshape([1,1,-1])).sum(axis = 2)
    if CI_type == "HPDI":
        interval = np.array([compute_hpdi(mu_test[:,i], credible_mass) for i in mu_test.shape[1]])
    else:
        interval = np.array([compute_quantile(mu_test[:,i], credible_mass) for i in mu_test.shape[1]])
    ret_df = pd.DataFrame.from_dict({
         "gene" : inference_res["hyperparams"]["gene_names"],
         "log_FC" : np.mean(mu_test, axis = 0),
         "CI_low" : interval[:,0], 
         "CI_high" : interval[:,1], 
         "is_zero_in" : (interval[:,0] < 0) & (interval[:,1] > 0) })
    
    
def compute_hpdi(samples, credible_mass=0.95):
    """
    Compute the HPDI of a given array of samples.

    Parameters
    ----------
    samples : numpy.ndarray
        A 1D array of samples.
    credible_mass : float, optional
        The desired mass of the HPDI interval (default is 0.95).

    Returns
    -------
    numpy.ndarray
        A 2-element array containing the lower and upper bounds of the HPDI.
    """
    sorted_samples = np.sort(samples)
    N = len(sorted_samples)
    n_samples_hpdi = int(np.floor(credible_mass * N))
    interval_widths = sorted_samples[n_samples_hpdi:] - sorted_samples[:-n_samples_hpdi]
    min_width_index = np.argmin(interval_widths)
    hpdi = np.array([sorted_samples[min_width_index],
                     sorted_samples[min_width_index + n_samples_hpdi]])

    return hpdi
    
    

def compute_quantile(samples, credible_mass=0.95):
    """
    Compute the quantile credible interval of a given array of samples.

    Parameters
    ----------
    samples : numpy.ndarray
        A 1D array of samples.
    credible_mass : float, optional
        The desired mass of the credible interval (default is 0.95).

    Returns
    -------
    numpy.ndarray
        A 2-element array containing the lower and upper bounds of the credible interval.
    """
    lower = (1 - credible_mass) / 2 * 100
    upper = (1 + credible_mass) / 2 * 100
    ci = np.percentile(samples, [lower, upper])

    return ci
   

tests/test_pydevil.py
from pydevil import pydevil


src/pydevil/guide.py
import pandas as pd
import numpy as np
import torch

import pyro 
import pyro.distributions as dist
from torch.distributions import constraints

from pydevil.utils import init_beta, init_theta


def guide(input_matrix, 
          model_matrix, 
          UMI, 
          group_matrix = None, 
          gene_specific_model_tensor = None,
          kernel_input = None,
          full_cov = True,
          prior_loc = 10, 
          batch_size = 5120, 
          theta_bounds = (1e-6, 10000),
          init_loc = 0.1):
  
    n_cells = input_matrix.shape[0]
    n_genes = input_matrix.shape[1]
    n_features = model_matrix.shape[1]
    
    beta_estimate = init_beta(torch.log((input_matrix + 1e-5) / UMI.unsqueeze(1)), model_matrix)
    theta_estimate = init_theta(input_matrix * 1.)    
    beta_mean = pyro.param("beta_mean", beta_estimate, constraint=constraints.real)
            
    if kernel_input is not None:
        lengthscale_par = pyro.param("lengthscale_param", torch.ones(n_genes), constraint=constraints.positive)
        
                    
    if group_matrix is not None:
        n_groups = group_matrix.shape[1]
        if full_cov:
            zeta_loc = pyro.param("zeta_loc", (torch.eye(n_groups, n_groups).repeat([n_genes,1,1]) * init_loc), constraint=constraints.lower_cholesky)
        else:
            zeta_loc = pyro.param("zeta_loc", torch.ones(n_genes, n_groups) * init_loc, constraint=constraints.positive) 
               
    if full_cov:
        beta_loc = pyro.param("beta_loc", (torch.eye(n_features, n_features).repeat([n_genes,1,1]) * init_loc), constraint=constraints.lower_cholesky)
    else:
        beta_loc = pyro.param("beta_loc", torch.ones(n_genes, n_features) * init_loc, constraint=constraints.positive) 
    
    theta_p = pyro.param("theta_p", theta_estimate, constraint=constraints.positive) 
    with pyro.plate("genes", n_genes, dim = -1):
        pyro.sample("theta", dist.Delta(theta_p))
        
        if kernel_input is not None:
            lengthscale = pyro.sample("lengthscale", dist.Delta(lengthscale_par))
            pyro.sample("kernel_random_effect", dist.MultivariateNormal(torch.zeros(n_cells), covariance_matrix= kernel_input, validate_args=False)).reshape([n_cells,n_genes]) * lengthscale
      
        if group_matrix is not None:
            if full_cov:
                zeta = pyro.sample("zeta", dist.MultivariateNormal(torch.zeros(n_genes, n_groups), scale_tril=zeta_loc, validate_args=False))
            else:
                zeta = pyro.sample("zeta", dist.Normal(torch.zeros(n_genes, n_groups), zeta_loc).to_event(1))

        if full_cov:
            pyro.sample("beta", dist.MultivariateNormal(beta_mean.t(),scale_tril = beta_loc, validate_args=False))
        else:
            pyro.sample("beta", dist.Normal(beta_mean.t(), beta_loc).to_event(1))


        
def guide_mle(*args, **kargs):
  pass
        
        

src/pydevil/model.py
import pandas as pd
import numpy as np
import torch

import pyro 
import pyro.distributions as dist
from torch.distributions import constraints

def model(input_matrix, 
          model_matrix, 
          UMI, 
          group_matrix = None, 
          gene_specific_model_tensor = None,
          kernel_input = None,
          full_cov = True,
          prior_loc = 10, 
          batch_size = 5120, 
          theta_bounds = (1e-6, 1000000),
          init_loc = 0.1):
  
  n_cells = input_matrix.shape[0]
  n_genes = input_matrix.shape[1]
  n_features = model_matrix.shape[1]
  lengthscale = torch.ones(n_genes)

  
     
  
  with pyro.plate("genes", n_genes, dim = -1):
    theta = pyro.sample("theta", dist.Uniform(theta_bounds[0],theta_bounds[1]))
    beta_prior_mu = torch.zeros(n_features)

    if kernel_input is not None:
      lengthscale = pyro.sample("lengthscale", dist.Gamma(torch.ones(n_genes), torch.ones(n_genes)))
      kernel_mu = pyro.sample("kernel_random_effect", dist.MultivariateNormal(torch.zeros(n_cells), scale_tril=kernel_input, validate_args=False)).reshape([n_cells,n_genes]) * lengthscale

    
    if group_matrix is not None:
      n_groups = group_matrix.shape[1]
  ### This is one possible implementation as a hierarchical model ###
      if full_cov:
        zeta = pyro.sample("zeta", dist.MultivariateNormal(torch.zeros(n_genes, n_groups), scale_tril=torch.eye(n_groups, n_groups) * prior_loc, validate_args=False))
      else:
        zeta = pyro.sample("zeta", dist.Normal(torch.zeros(n_genes, n_groups), torch.ones(n_groups) * prior_loc).to_event(1))
    if gene_specific_model_tensor is not None:
        n_features_gs = gene_specific_model_tensor.shape[1]
        beta_prior_mu_gs = torch.zeros(n_features_gs)
        if full_cov:
            beta_gs = pyro.sample("beta_gs", dist.MultivariateNormal(beta_prior_mu_gs, scale_tril=torch.eye(n_features_gs, n_features_gs) * prior_loc, validate_args=False))
        else:
            beta_gs = pyro.sample("beta_gs", dist.Normal(beta_prior_mu_gs, torch.ones(n_features_gs) * prior_loc).to_event(1))
        
    if full_cov:
      beta = pyro.sample("beta", dist.MultivariateNormal(beta_prior_mu, scale_tril=torch.eye(n_features, n_features) * prior_loc, validate_args=False))
    else:
      beta = pyro.sample("beta", dist.Normal(beta_prior_mu, torch.ones(n_features) * prior_loc).to_event(1))

    
    with pyro.plate("data", n_cells, dim = -2):
        eta = torch.matmul(model_matrix, beta.T)  + torch.log(UMI).unsqueeze(1)
        if group_matrix is not None:
            eta_zeta = torch.matmul(group_matrix , zeta.T)
            eta = eta + eta_zeta
        if kernel_input is not None:
            eta = eta + kernel_mu
        if gene_specific_model_tensor is not None:
            eta = eta + eta_gene_specific

        # print(theta.max())
        # print(theta.min())
        pyro.sample("obs", dist.GammaPoisson(rate = torch.clamp(torch.exp(eta + torch.log(theta)),0, 1e9) ,
        concentration= torch.clamp(1/theta, 1e-9,1e9)), obs = input_matrix)
    

src/pydevil/utils.py
import torch
import numpy as np
import pandas as pd
from functools import reduce


def check_is_categorical(mat):
    return (torch.sum(mat == 0) + torch.sum(mat == 1)) == (mat.shape[0] * mat.shape[1])

def init_beta(X, model_matrix, approx_type = "auto", nsamples = 10000):
    if approx_type == "auto":
        if check_is_categorical(model_matrix):
            approx_type = "group_mean"
        else:
            approx_type = "batch_linear_regression"
    if approx_type == "group_mean":
        beta_init = ((X.t() @ model_matrix) / model_matrix.sum(0)).t()
    else:
        if nsamples < model_matrix.shape[0]:
            n_categorical = torch.where(torch.tensor([check_is_categorical(model_matrix[:,i].unsqueeze(1)) for i in range(model_matrix.shape[1])]))[0]
            if n_categorical.shape[0] == 0:
                indexes = np.random.randint(model_matrix.shape[0], size = nsamples)
            else:
                samples_per_cat = nsamples//n_categorical.shape[0]                
                n_categorical = n_categorical.tolist()
                indexes = []
                for i in n_categorical:
                    samples_i = samples_per_cat
                    if samples_per_cat > (model_matrix[:,i] == 1).sum().item():
                        samples_i = (model_matrix[:,i] == 1).sum().item()
                    av_samples = np.array(np.random.choice(torch.where(model_matrix[:,i] == 1)[0].detach().numpy(),size = samples_i, replace=False))
                    indexes.append(av_samples)
                indexes = np.concatenate(indexes)
            X_regression = X[indexes,:]
            model_matrix_sub = model_matrix[indexes,:]
        else:
            X_regression = X
            model_matrix_sub = model_matrix
        beta_init =  torch.linalg.solve(model_matrix_sub.t() @ model_matrix_sub, model_matrix_sub.t() @ X_regression,  left = True)
    return beta_init

def init_theta(X, min_t = 0.0001, max_t = 1e6):
    variance = torch.var(X, dim=0)
    mean = torch.mean(X, dim=0)
    to_return = mean**2 / (variance - mean) 
    to_return = torch.clamp(torch.tensor(min_t), torch.tensor(max_t),to_return)
    to_return[torch.isnan(to_return)] = max_t
    return to_return.abs()


def rename_for_volcano(df_markers):
    df_markers_new = df_markers.set_axis(["GENE", "EFFECTSIZE", "p_value", "P", "is_significant"], axis=1)
    df_markers_new["SNP"] = df_markers_new["GENE"]
    df_markers_new["P"] = df_markers_new["P"] + 1e-16
    return df_markers_new

def prepare_batch(input_matrix, model_matrix, UMI, group_matrix, gene_specific_model_tensor, kernel_input, batch_size = 5124):
    
    if batch_size >= model_matrix.shape[0]:
        return input_matrix, model_matrix, UMI, group_matrix, gene_specific_model_tensor, kernel_input
    
    idx = torch.randperm(model_matrix.shape[0])[:batch_size]
    
    input_matrix_batch = input_matrix[idx,:]
    model_matrix_batch = model_matrix[idx,:]
    UMI_batch =  UMI[idx]
    
    if group_matrix is not None:
        group_matrix_batch = group_matrix[idx,:]
    else:
        group_matrix_batch = group_matrix
    
    if gene_specific_model_tensor is not None:
        gene_specific_model_tensor_batch = gene_specific_model_tensor[idx,:,:]
    else:
        gene_specific_model_tensor_batch = gene_specific_model_tensor
    
    if kernel_input is not None:
        kernel_input_batch = kernel_input[idx,:]
        kernel_input_batch = kernel_input_batch[:,idx]
    else:
        kernel_input_batch = kernel_input
    
    return input_matrix_batch, model_matrix_batch, UMI_batch, group_matrix_batch, gene_specific_model_tensor_batch, kernel_input_batch


def from_CNA_and_clones_to_mmatrix(input_CNA, cells_to_clones, gene_coord):
    
    clones = input_CNA.keys()
    clonal_profiles = pd.concat([build_CNA_profile_aux(gene_coord, input_CNA, clone) for clone in clones])
    common_genes = reduce(np.intersect1d,[cp.index.values for cp in clonal_profiles])
    clonal_profiles = [cp.loc[common_genes.tolist()] for cp in clonal_profiles]
    clonal_profiles = pd.concat(clonal_profiles, axis = 1)
    clonal_profiles = clonal_profiles.T.rename(columns= {"gene" : "clone"} )
    clonal_profiles["clone"] = clonal_profiles.index
    df_return = clonal_profiles.join(cell_to_clone.set_index('clone'), on = "clone").drop(columns=['clone', 'cell'])
    return df_return


def build_CNA_profile_aux(genes_to_chr,clones_profile, clone):
    cnv_df = clones_profile[clone]
    # Initialize an empty dataframe to store the result
    result_df = pd.DataFrame()

    # Loop over each row (gene) in the genes dataframe
    for gene_index, gene_row in genes_to_chr.iterrows():
        # Extract the start and end position of the current gene
        gene_start = gene_row['from']
        gene_end = gene_row['to']
        gene_chr = gene_row['chr']

        # Select rows (segments) from the cnv dataframe where the segment overlaps with the current gene
        # A segment overlaps with the gene if the segment starts before the gene ends AND the segment ends after the gene starts
        overlap_df = cnv_df[(cnv_df['start'] <= gene_end) & (cnv_df['end'] >= gene_start) & (cnv_df['chr'] == gene_chr)]

        # If any overlapping segments were found, add them to the result dataframe
        if not overlap_df.empty:
            overlap_df = overlap_df.copy()  # To avoid SettingWithCopyWarning
            overlap_df['gene'] = gene_row['gene']
            result_df = pd.concat([result_df, overlap_df])
    print(result_df)   
    return result_df.loc[:,"integer_copy_number"].set_index(result_df["gene"]).rename(columns = {"integer_copy_number" : clone})

src/pydevil/interface.py
import pandas as pd
import numpy as np
import torch

import pyro 
import pyro.distributions as dist
from torch.distributions import constraints
from pyro.infer import SVI, Trace_ELBO, JitTrace_ELBO, TraceGraph_ELBO, NUTS, MCMC
from pyro.infer.autoguide import AutoMultivariateNormal

import seaborn as sns
import matplotlib.pyplot as plt

import sys
from tqdm import trange

from pydevil.model import model
from pydevil.guide import guide
from pydevil.utils import prepare_batch

from sklearn.metrics.pairwise import rbf_kernel



def run_SVDE(
    input_matrix,
    model_matrix, 
    ncounts, 
    group_matrix = None,
    gene_specific_model_tensor = None,
    kernel_input = None,
    gene_names = None,
    cell_names = None,
    optimizer_name = "ClippedAdam",
    steps = 100, 
    lr = 0.1,
    gamma_lr = 0.1,
    cuda = False,
    jit_compile = False,
    batch_size = 5120, 
    full_cov = True, 
    prior_loc = 0.1,
    theta_bounds = (0., 1e16),
    init_loc = 0.1
):
  
    if cuda and torch.cuda.is_available():
        torch.set_default_tensor_type('torch.cuda.FloatTensor')
    else:
        torch.set_default_tensor_type(t=torch.FloatTensor)

    if jit_compile and not cuda:
        pyro_loss = JitTrace_ELBO
    else:
        pyro_loss = Trace_ELBO
        
    if batch_size > input_matrix.shape[0]:
        batch_size = input_matrix.shape[0]

    if kernel_input is not None:
        kernel_input = rbf_kernel(kernel_input, gamma = 1.) + np.eye(kernel_input.shape[0]) * 0.1
        kernel_input = torch.tensor(kernel_input).float()
        
    lrd = gamma_lr ** (1 / steps)

    
    

    
    input_matrix, model_matrix, UMI = torch.tensor(input_matrix).int(), torch.tensor(model_matrix).float(), torch.tensor(ncounts).float()

    if group_matrix is not None:
        group_matrix = torch.tensor(group_matrix).float()

    if gene_specific_model_tensor is not None:
        gene_specific_model_tensor = torch.tensor(gene_specific_model_tensor).float()
  
    if optimizer_name == "SGD":
        optimizer = pyro.optim.SGD({"lr": lr})
    elif optimizer_name == "Adam":
        optimizer = pyro.optim.Adam({"lr": lr})
    else:
        optimizer = pyro.optim.ClippedAdam({"lr": lr, "lrd" : lrd})
            
    elbo_list = [] 

    pyro.clear_param_store()
    
    t = trange(steps, desc='Bar desc', leave = True)
        
    svi = pyro.infer.SVI(model, guide, optimizer, pyro.infer.TraceGraph_ELBO())

    for i in t:

        input_matrix_batch, model_matrix_batch, UMI_batch, group_matrix_batch, \
        gene_specific_model_tensor_batch, kernel_input_batch =  \
        prepare_batch(input_matrix, model_matrix, UMI, group_matrix,  \
        gene_specific_model_tensor, kernel_input, batch_size = batch_size)

        loss = svi.step(input_matrix_batch, model_matrix_batch, UMI_batch, group_matrix_batch,
        gene_specific_model_tensor_batch, kernel_input_batch, full_cov = full_cov, 
        prior_loc = prior_loc,
        theta_bounds = theta_bounds,
        init_loc = init_loc)       

        elbo_list.append(loss)
        t.set_description('ELBO: {:.5f}  '.format(loss))
        t.refresh()
    
    
    n_features = model_matrix.shape[1]
    coeff = pyro.param("beta_mean")
    overdispersion = pyro.param("theta_p")

    if full_cov and n_features > 1:
        loc = torch.bmm(pyro.param("beta_loc"),pyro.param("beta_loc").permute(0,2,1))
    else:
        loc = pyro.param("beta_loc")

    eta = torch.exp(torch.matmul(model_matrix, coeff) + torch.unsqueeze(torch.log(UMI), 1) )
    lk = dist.NegativeBinomial(logits = eta - torch.log(overdispersion) ,
        total_count= torch.clamp(overdispersion, 1e-9,1e9)).log_prob(input_matrix).sum(dim = 0)

    if cuda and torch.cuda.is_available(): 
        input_matrix = input_matrix.cpu().detach().numpy() 
        overdispersion = overdispersion.cpu().detach().numpy() 
        eta = eta.cpu().detach().numpy()
        coeff = coeff.cpu().detach().numpy()
        loc = loc.cpu().detach().numpy()
        lk = lk.cpu().detach().numpy()
    else:
        input_matrix = input_matrix.detach().numpy() 
        overdispersion = overdispersion.detach().numpy() 
        eta = eta.detach().numpy()
        coeff = coeff.detach().numpy()
        loc = loc.detach().numpy()
        lk = lk.detach().numpy()

    variance =  eta + eta**2 / overdispersion

    ret = {"loss" : elbo_list, "params" : {
        "theta" : overdispersion,
        "lk" : lk,
        "beta" : coeff,
        "eta" : eta,
        "variance" : loc
        }, "residuals" : (input_matrix - eta) / np.sqrt(variance),
        "hyperparams" : {"gene_names" : gene_names, "cell_names" : cell_names ,"model_matrix" : model_matrix}     }

    if group_matrix is not None:
        n_random = group_matrix.shape[1]
        if full_cov and n_random > 1:
            ret["params"]["random_effects_variance"] = torch.bmm(pyro.param("zeta_loc"),pyro.param("zeta_loc").permute(0,2,1))
        else:
            ret["params"]["random_effects_variance"] = pyro.param("zeta_loc")
        if cuda and torch.cuda.is_available():
            ret["params"]["random_effects_variance"] = ret["params"]["random_effects_variance"].cpu().detach().numpy()
        else:
            ret["params"]["random_effects_variance"] = ret["params"]["random_effects_variance"].detach().numpy()
            
    if kernel_input is not None:
        if cuda and torch.cuda.is_available():
            ret["params"]["lengthscale_kernel"] = pyro.param("lengthscale_param").cpu().detach().numpy()
        else:
            ret["params"]["lengthscale_kernel"] = pyro.param("lengthscale_param").detach().numpy()
        
    return ret
  
  
def run_HMC( input_matrix,
    model_matrix, 
    ncounts, 
    group_matrix = None,
    gene_specific_model_tensor = None,
    kernel_input = None,
    gene_names = None,
    cell_names = None,
    num_samples = 1000,
    num_chains = 4,
    warmup_steps = 200,
    cuda = False, 
    jit_compile = True,
    full_cov = True, 
    prior_loc = 0.1,
    theta_bounds = (0., 1e16),
    init_loc = 0.1):
              

    if cuda and torch.cuda.is_available():
        mp_context = "spawn"
        torch.set_default_tensor_type('torch.cuda.FloatTensor')
    else:
        mp_context = None
        torch.set_default_tensor_type(t=torch.FloatTensor)
  
    if jit_compile and not cuda:
        loss = JitTrace_ELBO
    else:
        loss = Trace_ELBO
    
    if jit_compile and not cuda:
        pyro_loss = JitTrace_ELBO
    else:
        pyro_loss = Trace_ELBO
        

    if kernel_input is not None:
        kernel_input = rbf_kernel(kernel_input, gamma = 1.) + np.eye(kernel_input.shape[0]) * 0.1
        kernel_input = torch.tensor(kernel_input).float()    
    
    input_matrix, model_matrix, UMI = torch.tensor(input_matrix).int(), torch.tensor(model_matrix).float(), torch.tensor(ncounts).float()

    if group_matrix is not None:
        group_matrix = torch.tensor(group_matrix).float()

    if gene_specific_model_tensor is not None:
        gene_specific_model_tensor = torch.tensor(gene_specific_model_tensor).float()
    
    input_matrix, model_matrix, UMI = torch.tensor(input_matrix).float(), torch.tensor(model_matrix).float(), torch.tensor(ncounts).float()

    nuts_kernel = NUTS(model)

    mcmc = MCMC(nuts_kernel, num_samples=num_samples, warmup_steps=warmup_steps, num_chains = num_chains,mp_context = mp_context)
    mcmc.run(input_matrix, model_matrix, UMI, group_matrix,
        gene_specific_model_tensor, kernel_input, full_cov = full_cov, 
        prior_loc = prior_loc,
        theta_bounds = theta_bounds,
        init_loc = init_loc)
    if cuda: 
        hmc_samples = {k: v.detach().cpu().numpy() for k, v in mcmc.get_samples().items()}
    else:
        hmc_samples = {k: v.detach().numpy() for k, v in mcmc.get_samples().items()}
    ret = {"params" :  hmc_samples , 
             "hyperparams" : {"gene_names" : gene_names, "cell_names" : cell_names ,"model_matrix" : model_matrix}}

    return  ret
